jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java

```diff
@@ -439,80 +444,122 @@ private void checkImageInTargetRegistry() {
+                        Verify.verifyNotNull(
+                                results
+                                    .builtImagesAndContainerConfigurationPushResults
+                                    .get()
+                                    .keySet()
+                                    .iterator()
+                                    .next())
+                            .get())
                     .call());
   }
 
   private void pushImages() {
     ProgressEventDispatcher.Factory childProgressDispatcherFactory =
         Verify.verifyNotNull(rootProgressDispatcher).newChildProducer();
 
-    results.buildResult =
+    results.buildResults =
         executorService.submit(
             () -> {
-              realizeFutures(results.baseImageLayerPushResults.get().get(0));
               realizeFutures(results.applicationLayerPushResults.get());
 
-              List<Future<BuildResult>> manifestPushResults =
-                  scheduleCallables(
-                      PushImageStep.makeList(
-                          buildContext,
-                          childProgressDispatcherFactory,
-                          results.targetRegistryClient.get(),
-                          Verify.verifyNotNull(
+              List<Future<BuildResult>> buildResults = new ArrayList<>();
+              for (Map.Entry<Future<Image>, Future<BlobDescriptor>> entry :
+                  results.builtImagesAndContainerConfigurationPushResults.get().entrySet()) {
+
+                Future<BuildResult> buildResult =
+                    executorService.submit(
+                        () -> {
+                          realizeFutures(
+                              Verify.verifyNotNull(
                                   results
-                                      .builtImagesAndContainerConfigurationPushResults
+                                      .builtImagesAndLayerPushResults
                                       .get()
-                                      .get(results.builtImages.get().get(0)))
-                              .get(),
-                          results.builtImages.get().get(0).get(),
-                          results.manifestCheckResult.get().isPresent()));
-              realizeFutures(manifestPushResults);
-              return manifestPushResults.isEmpty()
-                  ? new BuildResult(
-                      results.manifestCheckResult.get().get().getDigest(),
-                      Verify.verifyNotNull(
-                              results
-                                  .builtImagesAndContainerConfigurationPushResults
-                                  .get()
-                                  .get(results.builtImages.get().get(0)))
-                          .get()
-                          .getDigest())
-                  // Manifest pushers return the same BuildResult.
-                  : manifestPushResults.get(0).get();
+                                      .get(entry.getKey())));
```

Comment:
@chanseokoh This is the line that has caused all the changes in the structure of the code .Do we really need to `realizeFutures(baseImageLayerPushResults) ` here ?

Is there any way we could for-go realizing these futures here? 

Once we have answered this question ,I can easily refactor the code.